### PR TITLE
add a rust GitHub action and cross test [skip ci]

### PR DESCRIPTION
**Pull request by Jimexist**
- closes #2594 

as it replaces both

- [x] Did you create an [Apache Jira](https://issues.apache.org/jira/projects/THRIFT/issues/) ticket?  (not required for trivial changes)
- [x] If a ticket exists: Does your pull request title follow the pattern "THRIFT-NNNN: describe my issue"?
- [x] Did you squash your changes to a single commit?  (not required, but preferred)
- [x] Did you do your best to avoid breaking changes?  If one was needed, did you label the Jira ticket with "Breaking-Change"?
- [x] If your change does not involve any code, include `[skip ci]` anywhere in the commit message to free up build resources.

### CODE REVIEW

File: lib/rs/src/protocol/compact.rs [hunk 1]
```diff
@@ -83,15 +83,13 @@ where
     fn read_list_set_begin(&mut self) -> crate::Result<(TType, i32)> {
         let header = self.read_byte()?;
         let element_type = collection_u8_to_type(header & 0x0F)?;
-
-        let element_count;
         let possible_element_count = (header & 0xF0) >> 4;
-        if possible_element_count != 15 {
+        let element_count = if possible_element_count != 15 {
```
Comment (allengeorge) [inline lib/rs/src/protocol/compact.rs:87]: Oh - didn't know you could do that now.

File: test/rs/src/bin/test_client.rs [hunk 2]
```diff
@@ -492,8 +483,8 @@ fn make_thrift_calls(
         s_cmp_nested_2.insert(Numberz::SIX, empty_insanity);
 
         let mut s_cmp: BTreeMap<UserId, BTreeMap<Numberz, Insanity>> = BTreeMap::new();
-        s_cmp.insert(1 as UserId, s_cmp_nested_1);
-        s_cmp.insert(2 as UserId, s_cmp_nested_2);
+        s_cmp.insert(1_i64, s_cmp_nested_1);
```
Comment (allengeorge) [inline test/rs/src/bin/test_client.rs:486]: Could you leave these two as before please?

File: test/rs/src/bin/test_client.rs [hunk 3]
```diff
@@ -492,8 +483,8 @@ fn make_thrift_calls(
...
+        s_cmp.insert(2_i64, s_cmp_nested_2);
```
Comment (allengeorge) [inline test/rs/src/bin/test_client.rs:487]: And this?

File: lib/rs/src/server/multiplexed.rs [hunk 4]
```diff
@@ -272,10 +271,10 @@ mod tests {
             _: &mut dyn TInputProtocol,
             _: &mut dyn TOutputProtocol,
         ) -> crate::Result<()> {
-            let res = self
-                .invoked
-                .compare_and_swap(false, true, Ordering::Relaxed);
-            if res {
+            let res =
```
Comment (allengeorge) [inline lib/rs/src/server/multiplexed.rs:274]: Interesting: I had no idea this was deprecated now.

File: lib/rs/test/src/bin/kitchen_sink_client.rs [hunk 5]
```diff
@@ -33,7 +33,6 @@ use kitchen_sink::recursive;
 use kitchen_sink::recursive::{CoRec, CoRec2, RecList, RecTree, TTestServiceSyncClient};
 use kitchen_sink::ultimate::{FullMealServiceSyncClient, TFullMealServiceSyncClient};
 
-use thrift;
```
Comment (allengeorge) [inline lib/rs/test/src/bin/kitchen_sink_client.rs:36]: Is there a reason we don't have to include the `use` definition here?

File: test/rs/src/bin/test_server.rs [hunk 6]
```diff
@@ -16,14 +16,12 @@
 // under the License.
 
 use clap::{clap_app, value_t};
-use env_logger;
```
Comment (allengeorge) [inline test/rs/src/bin/test_server.rs:19]: I'm a little surprised that we don't need the use declaration anymore.

File: test/rs/src/lib.rs [hunk 7]
```diff
@@ -15,5 +15,11 @@
 // specific language governing permissions and limitations
 // under the License.
 
+// FIXME - need changes in gen before lifting this exception
+#![allow(
```
Comment (allengeorge) [inline test/rs/src/lib.rs:19]: Could you create a followup JIRA for this?

**Review by allengeorge — COMMENTED** [review]
Reviewed everything but the github actions. In general this looks great! Were you able to run the cross tests successfully?

**Comment by allengeorge** [comment]
Unfortunately it looks like all the rust cross-tests are failing in the "Build/cross-test" action - even the rust<->rust ones, which should always work.

File: test/rs/src/bin/test_client.rs (hunk 2 above)
Comment (OmmyZhang) [reply test/rs/src/bin/test_client.rs:486]: `1 as UserId` will cause clippy::unnecessary_cast

```
error: casting integer literal to `i64` is unnecessary
```

File: test/rs/src/bin/test_server.rs (hunk 6 above)
Comment (OmmyZhang) [reply test/rs/src/bin/test_server.rs:19]: @allengeorge this is clippy::single_component_path_imports

```
error: this import is redundant
```

**Comment by Jimexist** [comment]
i'm merging this given the failure is due to missing uuid impl in rust:

```
testUuid("00112233-4455-6677-8899-aabbccddeeff")
*** FAILURE ***
org.apache.thrift.TApplicationException: unknown method testUuid
	at org.apache.thrift.TServiceClient.receiveBase(TServiceClient.java:81)
	at thrift.test.ThriftTest$Client.recv_testUuid(ThriftTest.java:527)
	at thrift.test.ThriftTest$Client.testUuid(ThriftTest.java:514)
	at org.apache.thrift.test.TestClient.main(TestClient.java:257)
Min time: 0us
Max time: 0us
Avg time: 0us
```

**Pull request merged**